File: day_02/python/day02.py
import string 


def main():
    file_data = open(r"../input.txt","r")
    ids = file_data.read().split("\n")
    print(part_one(ids))
    print(part_two(ids))


def part_one(ids):
    repeat_2 = set()
    repeat_3 = set()
    for id in ids:
        for letter in string.ascii_lowercase:
            result = id.count(letter)
            if result is 2:
                repeat_2.add(id)
            if result is 3:
                repeat_3.add(id)
    for q in repeat_3:
        print(q)
    return len(repeat_2) * len(repeat_3) 


def part_two(ids):
    target_length = len(ids[0])-1
    current_ids = ids.copy()
    for current_id in current_ids:
        ids.remove(current_id)
        for id in ids:
            return_value = ""
            for i in range(0, len(id)):
                if current_id[i] is id[i]:
                    return_value = return_value + current_id[i]
                if (len(return_value) is target_length):
                    return return_value


if __name__ == "__main__": 
    main()

File: day_02/java/day02.java
package advent2018.day_02.java;

import java.io.*; 
import java.util.*; 

public class day02 {
    public static void main(String[] args) throws Exception {
        String filePath = new File("").getAbsolutePath();
        Scanner inFile = new Scanner(new File(filePath + "/day_02/input.txt")).useDelimiter("\n");
        List<String> ids = new ArrayList<String>();
        while (inFile.hasNext()) {
            ids.add(inFile.next());
        }

        System.out.println(partOne(ids));
        System.out.println(partTwo(ids));
    }

    public static int partOne(List<String> ids) {
        Set<String> repeat2 = new HashSet<String>();
        Set<String> repeat3 = new HashSet<String>();
        char[] alphabet = "abcdefghijklmnopqrstuvwxyz".toCharArray();
        for (String id: ids) {
            for (char letter: alphabet) {
                String tempId = id + " ";
                int result = tempId.split(String.valueOf(letter)).length-1;
                if (result == 2) {
                    repeat2.add(id);
                }
                if (result == 3) {
                    repeat3.add(id);
                }
            }
        }
        return repeat2.size() * repeat3.size();
    }

    public static String partTwo(List<String> ids) {
        String s1, s2, returnValue = "";
        Integer targetLength = ids.get(0).length()-1;
        
        for (int i = 0; i < ids.size(); i++) {
            for (int j = i+1; j < ids.size(); j++) {
                s1 = ids.get(i);
                s2 = ids.get(j);
                returnValue = "";
                for (int k = 0, n = s1.length(); k < n; k++) {
                    if (s1.charAt(k) == s2.charAt(k)) {
                        returnValue += s1.charAt(k);
                    }
                }
                if (returnValue.length() == targetLength) {
                    return returnValue;
                }
            }
        }
        return ""; // should never ever happen
    }

}